HeterogeneousCore/AlpakaInterface/interface/prefixScan.h

```diff
@@ -0,0 +1,223 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_prefixScan_h
+#define HeterogeneousCore_AlpakaInterface_interface_prefixScan_h
+// Active version
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
+namespace cms {
+  namespace alpakatools {
+
+    template <typename TAcc, typename T, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE void warpPrefixScan(const TAcc& acc, int32_t laneId, T const* ci, T* co, uint32_t i) {
+      // ci and co may be the same
+      auto x = ci[i];
+      CMS_UNROLL_LOOP
+      for (int32_t offset = 1; offset < alpaka::warp::getSize(acc); offset <<= 1) {
+        // Force the exact type for integer types otherwise the compiler will find the template resolution ambiguous.
+        using dataType = std::conditional_t<std::is_floating_point_v<T>, T, std::int32_t>;
+        T y = alpaka::warp::shfl(acc, static_cast<dataType>(x), laneId - offset);
+        if (laneId >= offset)
+          x += y;
+      }
+      co[i] = x;
+    }
+
+    template <typename TAcc, typename T, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE void warpPrefixScan(const TAcc& acc, int32_t laneId, T* c, uint32_t i) {
+      warpPrefixScan(acc, laneId, c, c, i);
+    }
+
+    // limited to warpSize² elements
+    template <typename TAcc, typename T>
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE void blockPrefixScan(
+        const TAcc& acc, T const* ci, T* co, int32_t size, T* ws = nullptr) {
+      if constexpr (!requires_single_thread_per_block_v<TAcc>) {
+        const auto warpSize = alpaka::warp::getSize(acc);
+        int32_t const blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        int32_t const blockThreadIdx(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        ALPAKA_ASSERT_OFFLOAD(ws);
+        ALPAKA_ASSERT_OFFLOAD(size <= alpaka::warp::getSize(acc) * warpSize);
+        ALPAKA_ASSERT_OFFLOAD(0 == blockDimension % warpSize);
+        auto first = blockThreadIdx;
+        auto laneId = blockThreadIdx & (warpSize - 1);
```

Comment:
This requires that `warpSize` be a power of 2.
So far, all implementations I know of meet this requirement.
Would it make sense to check that ?

Reply:
Or change it to
```suggestion
        auto laneId = blockThreadIdx % warpSize;
```
?

Reply:
Cannot be static, of course, so made it a runtime (but debug version only) assertion.